rllib/offline/offline_prelearner.py

```diff
@@ -526,20 +508,64 @@ def _map_sample_batch_to_episode(
                 NotImplementedError
             else:
                 # Unpack observations, if needed.
-                obs = (
-                    unpack_if_needed(obs.tolist())
-                    if schema[Columns.OBS] in input_compress_columns
-                    else obs.tolist()
-                )
-                # Append the last `new_obs` to get the correct length of observations.
-                obs.append(
-                    unpack_if_needed(batch[schema[Columns.NEXT_OBS]][i][-1])
-                    if schema[Columns.OBS] in input_compress_columns
-                    else batch[schema[Columns.NEXT_OBS]][i][-1]
-                )
+                if isinstance(obs, str):
+                    # Decompress the observations
+                    obs = unpack_if_needed(obs)
+                    # Convert to a list of arrays.
+                    obs = [obs[i, ...] for i in range(obs.shape[0])]
+                else:
+                    obs = (
+                        unpack_if_needed(obs.tolist())
+                        if schema[Columns.OBS] in input_compress_columns
+                        else obs.tolist()
+                    )
+
+                if schema[Columns.NEXT_OBS] in batch:
+                    # Append the last `new_obs` to get the correct length of
+                    # observations.
+                    obs.append(
+                        unpack_if_needed(batch[schema[Columns.NEXT_OBS]][i][-1])
+                        if schema[Columns.OBS] in input_compress_columns
+                        else batch[schema[Columns.NEXT_OBS]][i][-1]
+                    )
+                else:
+                    # Otherwise we duplicate the last observation.
+                    obs.append(obs[-1])
+
+                # Check, if we have `done`, `truncated`, or `terminated`s in
+                # the batch.
+                if (
+                    schema[Columns.TRUNCATEDS] in batch
+                    and schema[Columns.TERMINATEDS] in batch
+                ):
+                    truncateds = batch[schema[Columns.TRUNCATEDS]][i][-1]
+                    terminateds = batch[schema[Columns.TERMINATEDS]][i][-1]
+                elif (
+                    schema[Columns.TRUNCATEDS] in batch
+                    and schema[Columns.TERMINATEDS] not in batch
+                ):
+                    truncateds = batch[schema[Columns.TRUNCATEDS]][i][-1]
+                    terminateds = False
+                elif (
+                    schema[Columns.TRUNCATEDS] not in batch
+                    and schema[Columns.TERMINATEDS] in batch
+                ):
+                    terminateds = batch[schema[Columns.TERMINATEDS]][i][-1]
+                    truncateds = False
+                elif "done" in batch:
+                    terminateds = batch["done"][i][-1]
+                    truncateds = False
+                else:
```

Comment:
add one-line comment: If terminated/truncated/done entirely missing, assume the episode chunk is terminated.

Reply:
What do you think? Maybe the default should be truncated=True (instead of terminated=True) for this case? 🤔

Reply:
I am on your side here, but the customer has completely terminated episodes in their recorded `SampleBatch`es and no `terminated`, `truncated` nor `done` in the batch. I keep a `TODO` for us to switch to `truncated=True` as soon as this is not needed anymore.

Reply:
Also: If the episodes are not terminated, we get problems in the replay buffers as they try to concatenate episode chunks with the same ID and can do so only in timely order, which cannot be guaranteed when sampling from a dataset - specifically not from a TB-sized one.